spot_ros2_control/hardware/src/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,354 @@
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 1]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 2]));
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface> SpotHardware::export_command_interfaces() {
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto& joint = info_.joints.at(i);
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                         &hw_commands_[interfaces_per_joint_ * i]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 1]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 2]));
+  }
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_activate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // command and state should be equal when starting
+  hw_commands_.assign(hw_states_.begin(), hw_states_.end());
+  // Set up the robot using the BD SDK.
+  if (!authenticate_robot(hostname_, username_, password_)) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_time_sync()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!check_estop()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!get_lease()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!power_on()) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+  if (!start_state_stream(std::bind(&StateStreamingHandler::handle_state_streaming, &state_streaming_handler_,
+                                    std::placeholders::_1))) {
+    return hardware_interface::CallbackReturn::ERROR;
+  }
+
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_deactivate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  stop_state_stream();
+  release_lease();
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_shutdown(const rclcpp_lifecycle::State& /*previous_state*/) {
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
```

Comment:
@khughes-bdai this may be a little trickier than it seems. System may shutdown while active too, without going through deactivation. I think we have to ensure cleanup is idempotent (ie. it's a no-op if there's nothing to clean up).